Minor array block cleanup

alex-spies opened the pull request:
While working on `mv_expand`, I came upon a couple minor things that I think are worth cleaning up:
- During tests, ensure that an `X-ArrayBlock`'s (and `X-BigArrayBlock`'s) invariants are actually valid; these invariants are currently only implicit and not documented.
- Simplify the initial allocation of the shifted `nullsMask` when expanding; currently we compute the number of non-null values, which does not indicate how big the shifted nulls mask should be - just use the current nulls mask as lower bound instead, avoiding unnecessary computations.
- Estimate the size of a BitSet more precisely to avoid (minor) underestimations, esp. when expanding.

alex-spies committed:
Enforce invariants in array blocks when testing

alex-spies committed:
Estimate BitSet ram usage more precisely

alex-spies committed:
Avoid computing nulls when shifting in expand

Computing a block's total value count to get an estimate for the
expanded nullsMask's size is unnecessary. We obtain a proper lower bound
by using the unexpanded nullsMask already.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

dnhatn commented on x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java:
```diff
@@ -50,6 +50,23 @@ private BooleanArrayBlock(
     ) {
         super(positionCount, firstValueIndexes, nulls, mvOrdering, blockFactory);
         this.vector = vector;
+        checkInvariants();
```
I think we should name this method `assertInvariants` and return a boolean, then call `assert checkInvariants()` instead.

nik9000 commented on x-pack/plugin/esql/compute/src/main/generated-src/org/elasticsearch/compute/data/BooleanArrayBlock.java:
```diff
@@ -50,6 +50,23 @@ private BooleanArrayBlock(
+        checkInvariants();
+    }
+
+    private void checkInvariants() {
```
To make it a little more obvious this is "assertion only" code, you could make this return a dummy boolean you can `assert` it. Or you could use `if (Assertions.ENABLED == false) {return}` at the front of it.

dnhatn reviewed (approved):
One comment, but LGTM. Thanks, Alex!

alex-spies committed:
Make it obvious that invariant check only asserts

alex-spies committed:
Move invariant check into AbstractBlock

alex-spies committed:
Make invariants slightly stricter

In particular, ensure no out of bounds access to the firstValueIndexes
array while expanding by asserting the nullsMask's last set bit.

alex-spies committed:
Avoid underestimating nullsMask size in expand

Sometimes, the nullsMask is actually longer than necessary; since
shiftNullsToExpandPositions initially allocates a nullsMask of the same
size, this should be reflected in the estimate for the expanded
nullsMask's size.

alex-spies committed:
Spotless

alex-spies commented:
Thanks for the reviews @dnhatn and @nik9000 !

I implemented your remarks. Additionally, I made the invariant check slightly stricter and the `nullsMask` estimate when expanding slightly more precise.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/AbstractBlock.java:
```diff
@@ -43,6 +44,26 @@ protected AbstractBlock(int positionCount, @Nullable int[] firstValueIndexes, @N
+            }
+        }
+        if (nullsMask != null) {
+            assert nullsMask.nextSetBit(getPositionCount() + 1) == -1;
```
This is new. This was assumed in `shiftNullsToExpandedPositions` may end up accessing `firstValueIndexes` out of bounds.

alex-spies commented on x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/data/X-BigArrayBlock.java.st:
```diff
@@ -106,7 +109,7 @@ public final class $Type$BigArrayBlock extends AbstractArrayBlock implements $Ty
         // The following line is correct because positions with multi-values are never null.
         int expandedPositionCount = vector.getPositionCount();
-        long bitSetRamUsedEstimate = BlockRamUsageEstimator.sizeOfBitSet(expandedPositionCount);
+        long bitSetRamUsedEstimate = Math.max(nullsMask.size(), BlockRamUsageEstimator.sizeOfBitSet(expandedPositionCount));
```
This is also new; I realized that sometimes the initial `nullsMask` is longer than necessary (even taking into account that `nullsMask.size()` is a multiple of `Long.SIZE`). Since `shiftNullsToExpandedPositions` now just pre-allocates as much memory for the expanded `nullsMask` as for the initial `nullsMask`, this is slightly more correct.
